homeassistant/components/vacuum/mqtt.py

```diff
@@ -0,0 +1,251 @@
+"""
+Support for a generic MQTT vacuum.
+
+For more details about this platform, please refer to the documentation
+https://home-assistant.io/components/vacuum/
+"""
+import asyncio
+import json
+import logging
+
+import voluptuous as vol
+
+import homeassistant.components.mqtt as mqtt
+import homeassistant.helpers.config_validation as cv
+from homeassistant.components.vacuum import (
+    DEFAULT_ICON, SUPPORT_BATTERY, SUPPORT_CLEAN_SPOT,
+    SUPPORT_LOCATE, SUPPORT_PAUSE, SUPPORT_RETURN_HOME,
+    SUPPORT_STATUS, SUPPORT_STOP, SUPPORT_TURN_OFF,
+    SUPPORT_TURN_ON, VacuumDevice, services_to_strings, strings_to_services)
+from homeassistant.const import CONF_NAME, ATTR_SUPPORTED_FEATURES,\
+    ATTR_BATTERY_LEVEL, ATTR_STATE
+from homeassistant.core import callback
+
+_LOGGER = logging.getLogger(__name__)
+
+DEPENDENCIES = ['mqtt']
+
+DEFAULT_SERVICES = SUPPORT_TURN_ON | SUPPORT_TURN_OFF | SUPPORT_STOP |\
+                   SUPPORT_RETURN_HOME | SUPPORT_STATUS | SUPPORT_BATTERY |\
+                   SUPPORT_CLEAN_SPOT
+ALL_SERVICES = DEFAULT_SERVICES | SUPPORT_PAUSE | SUPPORT_LOCATE
+DEFAULT_SERVICE_STRINGS = services_to_strings(DEFAULT_SERVICES)
+
+DEFAULT_COMMAND_TOPIC = 'vacuum/command'
+DEFAULT_STATE_TOPIC = 'vacuum/state'
+
+DEFAULT_NAME = 'MQTT Vacuum'
+
+PLATFORM_SCHEMA = mqtt.MQTT_BASE_PLATFORM_SCHEMA.extend({
+    vol.Optional(ATTR_SUPPORTED_FEATURES, default=DEFAULT_SERVICE_STRINGS):
+        vol.All(cv.ensure_list, [cv.string]),
+    vol.Optional(mqtt.CONF_COMMAND_TOPIC, default=DEFAULT_COMMAND_TOPIC):
+        mqtt.valid_publish_topic,
+    vol.Optional(mqtt.CONF_STATE_TOPIC, default=DEFAULT_STATE_TOPIC):
+        mqtt.valid_publish_topic,
+    vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+})
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Set up the vacuum."""
+    name = config.get(CONF_NAME)
+    command_topic = config.get(mqtt.CONF_COMMAND_TOPIC)
+    state_topic = config.get(mqtt.CONF_STATE_TOPIC)
+    supported_feature_strings = config.get(ATTR_SUPPORTED_FEATURES)
+    supported_features = strings_to_services(supported_feature_strings)
+    qos = config.get(mqtt.CONF_QOS)
+    state_template = cv.template('{{value_json.battery_level}}')
+    state_template.hass = hass
+
+    add_devices([
+        MqttVacuum(name, supported_features, command_topic, state_topic, qos,
+                   state_template),
+    ])
+
+
+class MqttVacuum(VacuumDevice):
+    """Representation of a MQTT-controlled vacuum."""
+
+    # pylint: disable=no-self-use
+    def __init__(self, name, supported_features, command_topic, state_topic,
+                 qos, state_template):
+        """Initialize the vacuum."""
+        self._name = name
+        self._command_topic = command_topic
+        self._state_topic = state_topic
+        self._supported_features = supported_features
+        self._qos = qos
+
+        self._payload_turn_on = "turn_on"
```

Comment:
Other mqtt platforms have configurable payloads. Maybe we want that here too, continuing @pvizeli's suggestion below on generalization.

Reply:
Totally open to that. It will add a ton of configuration options which might look intimidating to newer users. But I think that's probably an ok tradeoff